### PR TITLE
Add betterC probing for performance and for applications without drun…

### PR DESCRIPTION
**Pull request by etcimon**
This improves performance and also allows me to compile a custom druntime with my application. (post-switches = [] in the ldc2.conf)

### CODE REVIEW

**Comment by rikkimax** [comment]
Duplicate of: https://github.com/dlang/dub/pull/2699

**Comment by etcimon** [comment]
My solution involves minor changes and reverts to old behaviour, easier to approve.

File: source/dub/compilers/compiler.d
Comment (thewilsonator) [inline source/dub/compilers/compiler.d:187]: `-betterC` is called something else with GDC ,`-f-no-druntime` or something like that.

Comment (etcimon) [reply source/dub/compilers/compiler.d]: Seems to be -fno-druntime

File: source/dub/compilers/utils.d
Comment (Geod24) [inline source/dub/compilers/utils.d:286]: I'm a bit concerned about this part, as it seems to be specific to the current state of BetterC. Future improvements might make this out of date. Am I right that the problem arises because the platform probe has a dependency on the runtime ? Couldn't we achieve something similar with a sufficiently sized static array ?

Comment (etcimon) [reply source/dub/compilers/utils.d:286]: It's the string[] appender running as CTFE that calls the runtime as a stub, if there's any improvements we would be removing some definitions and that's pretty much it. Either way if anything breaks it'll show a warning and fallback to using the runtime (through the bool retry in the catch)

Comment (etcimon) [reply source/dub/compilers/utils.d:286]: The CTFE doesn't support using static storage either, you have to use the GC to create an allocation and then you can reference it in scope

Comment (etcimon) [reply source/dub/compilers/utils.d]: I managed to remove the declarations by using only pragma(msg, ) and formatting the string at the end

**Comment by Geod24** [comment]
Just to be sure I follow: Do we need the `-betterC` flag if we make it `-betterC`-compliant ? As in, if you get it to a point it does not pull anything from the runtime, the flag becomes superfluous, doesn't it ?

**Comment by rikkimax** [comment]
> Just to be sure I follow: Do we need the `-betterC` flag if we make it `-betterC`-compliant ? As in, if you get it to a point it does not pull anything from the runtime, the flag becomes superfluous, doesn't it ?

No, it doesn't.

You will still be linking against druntime, and that has cost.

File: source/dub/compilers/compiler.d
Comment (etcimon) [reply source/dub/compilers/compiler.d:182]: I wasn't able to append commas properly in the pragmas so I thought sdl syntax is easier. This probe is basically an internal hack anyways.

**Comment by etcimon** [comment]
The `probePlatform` routine takes ~25ms (vs ~38ms previously) for me using a `StopWatch` when building with these improvements. It's now also compatible with projects that don't include druntime.

**Comment by etcimon** [comment]
On dmd removing the `-betterC` flag improves it to 18ms, but with ldc it worsens it to 40ms.

**Comment by Geod24** [comment]
> You will still be linking against druntime, and that has cost.

Why do we need to link ?

File: .gitignore
Comment (etcimon) [reply .gitignore:19]: It's a minor annoyance when using vscode source control on windows, I was thinking it didn't require mention, nor its own pull request.

**Comment by etcimon** [comment]
It wouldn't link, it doesn't write an object file, compile flag `-o-`. It's mostly dmd.dinterpret in the compiler doing most of the work and returning after.

**Comment by etcimon** [comment]
Originally it's because this project https://github.com/etcimon/libwasm implements its own druntime and for it you must remove the includes in ldc2.conf, but probing fails on that configuration if it depends on druntime.

**Comment by Geod24** [comment]
> but probing fails on that configuration if it depends on druntime.

I assume that probing fails because it tries to import druntime, rather than link it ?

**Comment by etcimon** [comment]
It complains that there is no `module object`

**Comment by Geod24** [comment]
@kinke : Any chance you could have a look at this ? I think the use case is good, but I would like to make sure it keeps on working with newer versions of DMD/LDC with minimal maintenance.

File: source/dub/compilers/compiler.d
Comment (Geod24) [reply source/dub/compilers/compiler.d:182]: Well just `pragma(msg, ",");` ? Or something like:
```
pragma(msg, `"key": "`, value, `",`);
```
?

Comment (etcimon) [reply source/dub/compilers/compiler.d:182]: The trailing comma is invalid json and it's quite complicated to sanitize it

Comment (Geod24) [reply source/dub/compilers/compiler.d:182]: Fair. I'd like to reduce the dependency on SDL and keep the JSON one, but I don't think you should be blocked on this.

Comment (etcimon) [reply source/dub/compilers/compiler.d:182]: My solution would involve having invalid json spit out by the probe and then replacing `,]` with `]` afterwards, if that looks fine to you I rewrite it with json, but honestly even having a csv-like format and no json would make more sense.

**Comment by etcimon** [comment]
> So why is `-betterC` relevant here ?

It works without `-betterC`, but it does help with both performance and guarantees.

File: source/dub/compilers/compiler.d
Comment (Geod24) [reply source/dub/compilers/compiler.d:182]: Leave it as is, if I care enough I will make it use JSON5, but as you said, it's internal, so we can change it anytime.

**Comment by kinke** [comment]
AFAICT after a quick glance, the main change here is that the generated probe module becomes the special `object` module, only defining the `string` alias (not even `size_t`), so that the (fat) druntime one isn't imported. `-betterC` then shouldn't really make a difference, as its effects should be enabled automatically without `TypeInfo` class etc. in `object`.

While such a minimalistic, fully self-contained D module is surely fast to analyze with `-o-` and working for ~all exotic targets too, the cost is the ugly duplication in `source/dub/platform.d`. I guess that could be improved though by post-processing the existing string, e.g., replacing `ret ~= <literal>;` by `pragma(msg, ´ <literal>´);` and stripping off `string[] ret;` and `return ret;`.

**Comment by etcimon** [comment]
> I guess that could be improved though by post-processing the existing string

The post-processing using regex, matchAll and replace makes it impossible to evaluate the string with ctfe, but I doubt this would make it slower.

File: .gitignore
Comment (Geod24) [reply .gitignore:19]: What artifacts does vscode generate ? We can keep it in this PR, I just could not tell where it's coming from.

File: source/dub/compilers/compiler.d
Comment (Geod24) [inline source/dub/compilers/compiler.d:193]: Merge those two lines ?

File: source/dub/compilers/compiler.d
Comment (Geod24) [inline source/dub/compilers/compiler.d:199]: Revert

File: source/dub/compilers/utils.d
Comment (Geod24) [inline source/dub/compilers/utils.d:271]: Revert

File: source/dub/compilers/utils.d
Comment (Geod24) [inline source/dub/compilers/utils.d:281]: I would also modify the comment so the next person knows the platform probe should not rely on the runtime to make your use case possible.

**Review by Geod24 — APPROVED** [review]
Looks pretty nice, just a few nits. I don't know if you want/can add a test to avoid a regression, I'm happy to merge as-is otherwise once the nits have been addressed.

File: .gitignore
Comment (etcimon) [reply .gitignore:19]: This is for the dub.exe and dub.pdb files that are generated when building under windows.

**Comment by etcimon** [comment]
Should I squash it?

**Pull request merged**
